algosdk/future/transaction.py

```diff
@@ -219,11 +219,18 @@ def undictify(d):
                 args.update(KeyregNonparticipatingTxn._undictify(d))
                 txn = KeyregNonparticipatingTxn(**args)
             else:
-                if (d["votekey"] is None and
-                    d["selkey"] is None and
-                    d["votefst"] is None and
-                    d["votelst"] is None and
-                    d["votekd"] is None):
+                if ("votekey" not in d and
```

Comment:
@jasonpaulos I had to add this section here in an `if`-`elif` block. There are two reasons:
1. I wrote a test `test_write_read_keyregofflinetxn`. This test initializes a `KeyregOfflineTxn` and then calls `transaction.write_to_file`. I then try `transaction.retrieve_from_file` and verify that the two objects are identical. In this instance, when `write_to_file` is called, the original `KeyregTxn` class method `dictify` is called, and it asserts that the dictionary is populated with the keys `selkey`, `votefst` ... `nonpart`, even if they are `None`. As such, empty values are written to the file as well (which arguably might be an inefficiency?)
2. On the other hand, for `test_keyreg_txn_offline`, which is the original test that we reverted to, it strips empty values and thus a decoded string without `votekey`, `selkey`, etc. is assumed to be a `KeyregOfflineTxn`.

Given this, I decided to first check that the parameters are __not__ in the dictionary, which would imply that it might come from a string generated from `msgpack_encode`. If that first check passes, we then check if these values are `None`. If that fails, then it has to be a `KeyregOnlineTxn`.

Reply:
A more elegant solution would be to refactor the `dictify` method in `KeyregTxn`, but I'm not sure if the team wants to keep that untouched. The change would involve only dictifying `selkey`, `votefst` etc. iff they won't be `None`, which I think we have some guarantees on. That way, the classes would implicitly match the behavior of `msgpack_encode`, which strips `None` values.

Reply:
Thank you for noticing this! In this case I believe the better solution would be to modify the `KeyregTxn.dictify` to only add the fields that are not None. This is more in line with the behavior of the other transaction types.

Then it should only be necessary to check if `votekey` and other parameters are not present in the dictionary in order to call `KeyregOfflineTxn._undictify`.

Reply:
I've made the recommended fixes! This works now, and is probably the best solution since `KeyregTxn` is a class that might be used in the wild. A nicer solution would be to write `dictify` in `KeyregOnlineTxn` and `KeyregOfflineTxn` and `KeyregNonparticipatingTxn` and convert `KeyregTxn` to be an internal class that cannot be instantiated.